### PR TITLE
Switch from Rainloop to SnappyMail

### PR DESCRIPTION
**Pull request by Diman0**
## What type of PR?

Feature

## What does this PR do?
As discussed in the project meeting (#1582), we decided we want to switch from Rainloop to an alternative. Rainloop has multiple open security issues which were not patched for a long time. 

We decided to switch to SnappyMail because it is more secure and based on RainLoop. This means that users using RainLoop will still have a webmail that looks familiar for them.

This PR replaces RainLoop with SnappyMail.

### Related issue(s)
- #2215 
- #1582

## Prerequisites
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [x] In case of feature or enhancement: documentation updated accordingly
- [x] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/workflow.html#changelog) entry file.

### CODE REVIEW

**Comment by Diman0** [comment]
bors try

**Comment by Diman0** [comment]
bors try

**Comment by ShadwDrgn** [comment]
Would love to see this merged because rainloop and roundcube are awful but snappymail is wonderful and actually has hotkeys :D sorry if this didn't contribute much.

**Comment by tete2soja** [comment]
In addition, a critic issue is publicly available and the Rainloop team didn't respond to it:

* https://github.com/RainLoop/rainloop-webmail/issues/2142
* https://blog.sonarsource.com/rainloop-emails-at-risk-due-to-code-flaw

Can be a new reason to switch to SnappyMail.

File: webmails/snappymail/config/nginx-snappymail.conf [hunk 1]
```diff
@@ -2,33 +2,42 @@ server {
...
     client_max_body_size {{ MESSAGE_SIZE_LIMIT|int + 8388608 }};
 
     location / {
-        try_files $uri /index.php?$query_string;
+        try_files $uri $uri/ /index.php$args;
     }
 
     location ~ \.php$ {
+        include /etc/nginx/fastcgi_params;
+
```
Comment (nextgens) [inline webmails/snappymail/config/nginx-snappymail.conf:26]: I think you need a try_files here... or a way to check that the file actually exists.

See https://www.nginx.com/resources/wiki/start/topics/examples/phpfcgi/#connecting-nginx-to-php-fpm or https://nginx.org/en/docs/http/ngx_http_core_module.html#try_files (with two different locations depending on whether it does or doesn't exist)

File: core/nginx/conf/nginx.conf [hunk 2]
```diff
@@ -190,6 +190,9 @@ http {
       {% endif %}
       {% if ADMIN == 'true' %}
        location {{ WEB_ADMIN }} {
+         {% if WEB_ADMIN != '/' %}
+         rewrite ^({{ WEB_ADMIN }})$ $1/ permanent;
```
Comment (nextgens) [inline core/nginx/conf/nginx.conf:194]: This seems unrelated... why is it required?

File: webmails/snappymail/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ENV SNAPPYMAIL_URL https://github.com/the-djmaze/snappymail/releases/download/v2.13.4/snappymail-2.13.4.zip
+
+RUN apk add --no-cache \
+      curl unzip \
+ && cd /var/www/webmail \
+ && curl -L -O ${SNAPPYMAIL_URL} \
+ && unzip -q *.zip \
+ && rm -f *.zip \
+ && find . -type d -exec chmod 755 {} \; \
+ && find . -type f -exec chmod 644 {} \; \
```
Comment (nextgens) [inline webmails/snappymail/Dockerfile:75]: The two lines could be simplified with: ``chmod -R u+w,a+rX /var/www/webmail``

File: webmails/snappymail/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+# Install Snappymail from source
+ENV SNAPPYMAIL_URL https://github.com/the-djmaze/snappymail/releases/download/v2.13.4/snappymail-2.13.4.zip
+
+RUN apk add --no-cache \
+      curl unzip \
+ && cd /var/www/webmail \
+ && curl -L -O ${SNAPPYMAIL_URL} \
+ && unzip -q *.zip \
+ && rm -f *.zip \
```
Comment (nextgens) [inline webmails/snappymail/Dockerfile:73]: I'd pipe the curl output into unzip here

**Review by nextgens — CHANGES_REQUESTED** [review]

File: webmails/snappymail/config/nginx-snappymail.conf [hunk 5]
```diff
@@ -38,4 +47,9 @@ server {
     location ^~ /data {
       deny all;
     }
+
+    location /health {
+        add_header Content-Type text/plain;
+        return 200;
```
Comment (nextgens) [inline webmails/snappymail/config/nginx-snappymail.conf:53]: I wonder if it's worth making it a PHP script to ensure that FPM is fine too.

**Review by nextgens — CHANGES_REQUESTED** [review]

File: webmails/snappymail/config/nginx-snappymail.conf (hunk 5 above)
Comment (ghostwheel42) [reply webmails/snappymail/config/nginx-snappymail.conf:53]: This would be good. You don't need a script though, you can use `ping.path` and `ping.response` of the fpm daemon.

File: webmails/snappymail/config/nginx-snappymail.conf (hunk 1 above)
Comment (Diman0) [reply webmails/snappymail/config/nginx-snappymail.conf:26]: Good catch. I will change this.

File: core/nginx/conf/nginx.conf (hunk 2 above)
Comment (Diman0) [reply core/nginx/conf/nginx.conf:194]: I don't remember. I think I had some issues with my reverse proxy and added this for testing purposes. I will remove it. It should not have been added.

File: webmails/snappymail/config/nginx-snappymail.conf (hunk 5 above)
Comment (Diman0) [reply webmails/snappymail/config/nginx-snappymail.conf:53]: Yes. With this we test both nginx and fpm-daemon. I will enable this and modify the health check in the Dockerfile.

**Comment by Diman0** [comment]
> What are the plans for having this? Any ETA in plan?

This week I will try finishing this PR. Assuming the reviewers have no new findings, it should be part of master within a week.

**Comment by Diman0** [comment]
bors try

File: webmails/snappymail/config/nginx-snappymail.conf [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        fastcgi_pass unix:/var/run/php7-fpm.sock;
+        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
+        {% if WEB_WEBMAIL == '/' %}
+        fastcgi_param SCRIPT_NAME $fastcgi_script_name;
+        {% else %}
+        fastcgi_param SCRIPT_NAME {{WEB_WEBMAIL}}/$fastcgi_script_name;
+        {% endif %}
+    }
+
+    location ~ /\.ht {
```
Comment (nextgens) [inline webmails/snappymail/config/nginx-snappymail.conf:45]: I'd rather we denied access to any hidden file here

**Review by nextgens — APPROVED** [review]
LGTM.

File: webmails/snappymail/config/nginx-snappymail.conf (hunk 6 above)
Comment (nextgens) [reply webmails/snappymail/config/nginx-snappymail.conf:45]: My understanding is that it shouldn't: in your example the dot is a parameter (after the '?').
What would be blocked is http://example.com/webmail/.example

Comment (Diman0) [reply webmails/snappymail/config/nginx-snappymail.conf]: I have made the propose change

**Review by nextgens — APPROVED** [review]

**Review by ghostwheel42 — APPROVED** [review]
LGTM.

I was urprised the WEBMAIL variable in mailu.env is in fact only a bool: the only tests against it are == or != "none". So WEBMAIL=yesplease will work, too.
